src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/persistence/SQLitePersistor.java

```diff
@@ -179,6 +221,426 @@ synchronized int insertRow(String tableName, List<Object> row) throws SQLExcepti
+        }
+        if (pair.setter == null) {
+          throw new NoSuchMethodException(getNoGetterSetterExist(clz, field, GetterOrSetter.SETTER));
+        }
+        fieldToGetterSetterMap.put(field, pair);
+        fieldNameToGetterSetterMap.put(field.getName(), pair);
+      }
+    }
+  }
+
+  private String getNoGetterSetterExist(Class<?> clz, java.lang.reflect.Field field, GetterOrSetter getterOrSetter) {
+    String type;
+    switch (getterOrSetter) {
+      case GETTER:
+        type = "getter";
+        break;
+      case SETTER:
+        type = "setter";
+        break;
+      default:
+        throw new IllegalArgumentException("Unrecognized type: " + getterOrSetter);
+    }
+
+    StringBuilder sb = new StringBuilder("Could not find '");
+    sb.append(type)
+        .append("' for the field '")
+        .append(field.getName())
+        .append("' of class '")
+        .append(clz.getName())
+        .append("'. Getters are expected to start with 'get' or 'is' and setters are expected to start with 'set' and they are required to"
+        + " end with the name of the field (case insensitive.)");
+    return sb.toString();
+  }
+
+  private <T> ColumnValuePair writeCollectionReferenceColumn(java.lang.reflect.Field field, Method getter, T obj)
+      throws InvocationTargetException, IllegalAccessException, SQLException, NoSuchMethodException {
+    ColumnValuePair columnValuePair = new ColumnValuePair();
+    String columnName = NESTED_OBJECT_COLUMN_PREFIX + field.getName();
+
+    Collection<?> collection = (Collection<?>) getter.getReturnType().cast(getter.invoke(obj));
+    Map<String, List<Integer>> nestedPrimaryKeys = new HashMap<>();
+    for (Object o: collection) {
+      String myActualType = o.getClass().getSimpleName();
+      nestedPrimaryKeys.putIfAbsent(myActualType, new ArrayList<>());
+
+      Class typeArgClass = getGenericParamTypeOfMethodReturn(getter);
+
+      int id = writeImplInner(typeArgClass.cast(o));
+      nestedPrimaryKeys.get(myActualType).add(id);
+    }
+    JsonArray json = new JsonArray();
+    // Create fields with the collectionReferenceType columns
+    for (Map.Entry<String, List<Integer>> colNameEntry: nestedPrimaryKeys.entrySet()) {
+      JsonObject jsonObject = new JsonObject();
+      JsonArray jsonArrayInner = new JsonArray();
+      colNameEntry.getValue().forEach(rowId -> jsonArrayInner.add(rowId));
+
+      jsonObject.addProperty(TABLE_NAME_JSON_KEY, colNameEntry.getKey());
+      jsonObject.add(ROW_IDS_JSON_KEY, jsonArrayInner);
+
+      json.add(jsonObject);
+    }
+    columnValuePair.field = (DSL.field(DSL.name(columnName), String.class));
+    columnValuePair.value = json.toString();
+    return columnValuePair;
+  }
+
+  private <T> int writeImplInner(T obj)
+      throws IllegalStateException, IllegalAccessException, InvocationTargetException, SQLException, NoSuchMethodException {
+    Class<?> clz = obj.getClass();
+    String tableName = getTableNameFromClassName(clz);
```

Comment:
I still think we should not use class name directly as the table name. 
For example, if we wants to persist rca flowunit of different types, how are we going to persist them into a same table (i.e. RCA table) ? 
This also applies to remediation actions, in that case all remediation actions (cache actions, queue actions) will be persisted into a single table

Or should we create a base class for persister to use in that case ?

Reply:
It may not be a good idea to persist two different FLowUnits directly into the same table. These classes can have different Fields or different number of fields, that would throw a DB exception. More checks we have to do, to convert the error messages into understandable ones.

If we at all want to persist two different FlowUnits into the same table, then we should create an Object in both FlowUnits, and that we should ask to persist. That way, at Java level, it will be guranteed to not throw DB exceptions because of schema mismatch.